### PR TITLE
Remove “in” expression workaround

### LINKED ISSUE #168: Add expression "in"
The core sdk implemented a new `in` expression in https://github.com/mapbox/mapbox-gl-native/pull/16162. This ticket will track the implementation of the corresponding iOS bindings.

For reference, this is the Android implementation: https://github.com/mapbox/mapbox-gl-native-android/pull/171

/cc @mapbox/maps-ios

### PR DESCRIPTION
**Pull request by 1ec5**
Removed the `match`-based workaround for `in` expressions in favor of real support for an `in` expression operator, including support for testing whether a string is a substring of another string.

* [x] Replace `match`-based workaround with `in` operator mappings
* [x] Test substring matching
* [x] Implement `ANY`/`SOME` and `NONE` with `=` and `ALL` with `!=`
* [x] Document `in` expressions in “Information for Style Authors” guide
* [x] Document comparison modifiers in “Predicates and Expressions” guide

Fixes #168. Depends on #181. Retarget and rebase this PR onto master before merging.

/cc @mapbox/maps-ios

### CODE REVIEW

**Comment by 1ec5** [comment]
I threw in some additional support for NSExpression syntactic sugar so developers don’t feel like they have to match our examples exactly:

* `ANY x = y` is logically equivalent to `x CONTAINS y` and `y IN x`. All three forms evaluate to an `in` expression.
* `ALL x != y` is logically equivalent to `NOT x CONTAINS y` and `NOT y IN x`. All three forms evaluate to an `in` expression inside a `!` expression.

**Comment by jmkiley** [comment]
I tested this PR out with point features with a `title` attribute. `title` had a string containing `feature` for all three, but I wasn't able to get any features with `[NSPredicate predicateWithFormat:@"title CONTAINS 'feature'"];`

<details><summary> Sample code </summary>

    MGLPointFeature *pointfeature1 = [[MGLPointFeature alloc] init];
    pointfeature1.coordinate = CLLocationCoordinate2DMake(0, 33);
    [pointfeature1 setAttributes:@{@"title": [NSString stringWithFormat:@"feature-%f", pointfeature1.coordinate.longitude]}];

    MGLPointFeature *pointfeature2 = [[MGLPointFeature alloc] init];
    pointfeature2.coordinate = CLLocationCoordinate2DMake(0, 66);
    [pointfeature2 setAttributes:@{@"title": [NSString stringWithFormat:@"feature"]}];

    MGLPointFeature *pointfeature3 = [[MGLPointFeature alloc] init];
    pointfeature3.coordinate = CLLocationCoordinate2DMake(0, 99);
    [pointfeature3 setAttributes:@{@"title": [NSString stringWithFormat:@"feature"]}];
                                               
    NSArray *points = @[pointfeature1, pointfeature2, pointfeature3];
    MGLShapeSource *source = [[MGLShapeSource alloc] initWithIdentifier:@"source" features:points options:nil];

    [style addSource:source];

    MGLCircleStyleLayer *layer = [[MGLCircleStyleLayer alloc] initWithIdentifier:@"layer" source:source];
    layer.predicate = [NSPredicate predicateWithFormat:@"title CONTAINS 'feature'"];
    layer.circleRadius = [NSExpression expressionForConstantValue:@40];
    [style addLayer:layer];

</details>

**Comment by jmkiley** [comment]
Noting that `CONTAINS` seems to work in combination with the `NOT` operator. Two circles displayed with `[NSPredicate predicateWithFormat:@"NOT title CONTAINS '33'"];`.

Also, using `CONTAINS` with an `NSExpression` works. For example, I saw one circle with `layer.circleOpacity = [NSExpression expressionWithFormat:@"TERNARY(title CONTAINS '33', 1, 0)"];`.

**Comment by 1ec5** [comment]
> I tested this PR out with point features with a `title` attribute. `title` had a string containing `feature` for all three, but I wasn't able to get any features with `[NSPredicate predicateWithFormat:@"title CONTAINS 'feature'"];`

mapbox/mapbox-gl-native#16262

File: platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm [hunk 1]
```diff
@@ -115,14 +161,8 @@ - (id)mgl_jsonExpressionObject {
...
-            
-            NSExpression *matchExpression = [NSExpression expressionForFunction:@"MGL_MATCH"
-                                                                      arguments:@[self.leftExpression,
-                                                                                  self.rightExpression,
-                                                                                  [NSExpression expressionForConstantValue:@YES],
-                                                                                  [NSExpression expressionForConstantValue:@NO]]];
-
-            return matchExpression.mgl_jsonExpressionObject;
+            op = @"in";
+            break;
```
Comment (fabian-guerra) [inline platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm]: 👏🏼 this is simple

**Comment by 1ec5** [comment]
>> I tested this PR out with point features with a `title` attribute. `title` had a string containing `feature` for all three, but I wasn't able to get any features with `[NSPredicate predicateWithFormat:@"title CONTAINS 'feature'"];`
> 
> Thanks for noticing that: mapbox/mapbox-gl-native#16262.

`in` filters have been fixed in mbgl, and #210 will pull in that fix. No additional changes will be needed at the SDK level to support `IN` in predicates, but I’d like to add unit tests that round-trip between NSPredicate and mbgl filters, which will depend on #210.

**Pull request merged**